fedimint-core/src/tiered_multi.rs

```diff
@@ -204,68 +163,12 @@ where
         d: &mut D,
         modules: &ModuleDecoderRegistry,
     ) -> Result<Self, DecodeError> {
-        Ok(Self(BTreeMap::consensus_decode_from_finite_reader(
+        Ok(Self(Tiered::consensus_decode_from_finite_reader(
             d, modules,
         )?))
     }
 }
 
-pub struct TieredMultiZip<'a, I, T>
-where
-    I: 'a,
-{
-    iters: Vec<I>,
-    _pd: PhantomData<&'a T>,
-}
-
-impl<'a, I, C> TieredMultiZip<'a, I, C> {
-    /// Creates a new `TieredMultiZip` Iterator from `Notes` iterators. These
-    /// have to be checked for structural equality! There also has to be at
-    /// least one iterator in the `iter` vector.
-    pub fn new(iters: Vec<I>) -> Self {
-        assert!(!iters.is_empty());
-
-        TieredMultiZip {
-            iters,
-            _pd: PhantomData,
-        }
-    }
-}
-
-impl<'a, I, C> Iterator for TieredMultiZip<'a, I, C>
-where
-    I: Iterator<Item = (Amount, C)>,
-{
-    type Item = (Amount, Vec<C>);
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let mut notes = Vec::with_capacity(self.iters.len());
-        let mut amount = None;
-        for iter in &mut self.iters {
-            match iter.next() {
-                Some((amt, note)) => {
-                    if let Some(amount) = amount {
-                        // This may fail if notes weren't tested for structural equality
-                        assert_eq!(amount, amt);
```

Comment:
Note for reviewers: This assert is effectively removed in `new_aggregate_from_tiered_iter()` (which replaces `TieredMultiZip`). Does anyone know if this is a problem anywhere where `new_aggregate_from_tiered_iter()` is called?

Reply:
I think we will see weird failure modes if the supplied iterators over tiered aren't structurally equivalent, i.e. don't contain the same amount tiers. 

I like having a second line of defense, we could achieve something similar here by doing:

```rust
assert!(
    tiered_multi.summary()
        .iter()
        .map(|(_tier, count)| count)
        .all_equal(),
    "The supplied Tiered structs were not structurally equal"
);
```

([`all_equal` coming from `itertools`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.all_equal))

Reply:
Done. I added a TODO mentioning that this only asserts the shape of the _output_, not the input. Technically a regression in the strictness of this assertion, but I don't think it will likely matter.

Reply:
Uff, you are right, I'm dumb :see_no_evil:

Reply:
At least now all or none of the operations following on the `TieredMulti` will fail, but this isn't what I had in mind :/ don't want to unnecessarily block this, it likely doesn't matter indeed.